pulsar-broker/src/test/java/org/apache/pulsar/client/api/FailoverSubscriptionTest.java

```diff
@@ -0,0 +1,90 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.mledger.impl.ManagedCursorImpl;
+import org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl;
+import org.apache.pulsar.broker.BrokerTestUtil;
+import org.apache.pulsar.broker.service.persistent.PersistentDispatcherSingleActiveConsumer;
+import org.apache.pulsar.broker.service.persistent.PersistentTopic;
+import org.apache.pulsar.client.impl.ConsumerImpl;
+import org.awaitility.Awaitility;
+import org.awaitility.reflect.WhiteboxImpl;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+@Slf4j
+@Test(groups = "broker-api")
+public class FailoverSubscriptionTest extends ProducerConsumerBase {
+    @BeforeClass(alwaysRun = true)
+    @Override
+    protected void setup() throws Exception {
+        super.internalSetup();
+        super.producerBaseSetup();
+    }
+
+    @AfterClass(alwaysRun = true)
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @Test(timeOut = 30_000, invocationCount = 5)
+    public void testWaitingCursorsCountAfterSwitchingActiveConsumers() throws Exception {
+        final String tp = BrokerTestUtil.newUniqueName("persistent://my-property/my-ns/tp");
+        final String subscription = "s1";
+        admin.topics().createNonPartitionedTopic(tp);
+        admin.topics().createSubscription(tp, subscription, MessageId.earliest);
+        PersistentTopic topic = (PersistentTopic) pulsar.getBrokerService().getTopic(tp, false).join().get();
+        Map<String, ConsumerImpl<byte[]>> consumerMap = new HashMap<>();
+        ConsumerImpl<byte[]> firstConsumer = (ConsumerImpl<byte[]>) pulsarClient.newConsumer().topic(tp)
+                .subscriptionType(SubscriptionType.Failover).subscriptionName(subscription).subscribe();
+        consumerMap.put(firstConsumer.getConsumerName(), firstConsumer);
+        PersistentDispatcherSingleActiveConsumer dispatcher =
+                (PersistentDispatcherSingleActiveConsumer) topic.getSubscription(subscription).getDispatcher();;
+        for (int i = 0; i < 100; i++) {
+            ConsumerImpl<byte[]> consumerLoop = (ConsumerImpl<byte[]>) pulsarClient.newConsumer().topic(tp)
+                    .subscriptionType(SubscriptionType.Failover).subscriptionName(subscription).subscribe();
+            consumerMap.put(consumerLoop.getConsumerName(), consumerLoop);
+            if (dispatcher != null && dispatcher.getActiveConsumer() != null) {
+                String activeConsumer = dispatcher.getActiveConsumer().consumerName();
+                consumerMap.remove(activeConsumer).close();
+            }
+        }
+
+        ManagedLedgerImpl managedLedger = (ManagedLedgerImpl) topic.getManagedLedger();
+        ConcurrentLinkedQueue<ManagedCursorImpl> waitingCursors =
+                WhiteboxImpl.getInternalState(managedLedger, "waitingCursors");
```

Comment:
Please avoid using reflection in tests whenever possible. It makes code hard to refactor. Just add a getter with `@VisibleForTesting`.

Reply:
Addressed